Preparations to enable automatic tests for wxQt

AliKet opened the pull request.

Commit:
Don't override Set{Min,Max}Size() in wxWindowQt

This partially reverts commit 3c68696 (Make wxTopLevelWindow::SetSizeHints()
work in wxQt) because it breaks some wxSizer tests. and more importantly, it
breaks the documented behaviour of SetMinSize() which says:

Notice that calling this method doesn't prevent the program from making the
window explicitly smaller than the specified size by calling SetSize(), it just
ensures that it won't become smaller than this size during the automatic layout.

Commit:
Remove newly added functions from wxQt

QtSet{Min,Max}Size() were added for use by Set{Min,Max}Size() and
DoSetSizeHints() overrides. and since we have canceled overriding
Set{Min,Max}Size() in previous commit, there is no need to keep them
around without good reason.

Commit:
Enable automatic tests for wxQt running on GitHub

Also build without wxUIActionSimulator support for now as it is currently
broken under wxQt and causes the tests to instantly crash if enabled.

Commit:
Make sure a generic window (e.g. wxPanel) is created with a valid size

Such windows will not be shown and are ignored by Qt if added directly as children
to another window not managed by any sizer.

Commit:
Fix wxWindow::Update() under wxQt

Calling this function should immediately repaints the invalidated area of the window
which is the documented and expected behaviour.

Commit:
Make propagation test pass under wxQt

By waiting for the window to be ready before generating wxPaintEvent

AliKet commented:
I am already aware of the reported failures and will be corrected in the upcoming commits

Commit:
Make wxTextEntry::DoSetValue() reusable under wxQt

This is a preparation for wxComboBox and wxBitmapComboBox to reuse the code
from the base class by just forwarding to wxTextEntry.

Commit:
wxComboBox now forwards to wxTextEntry when setting its value under wxQt

Clean up no longer used code.

Commit:
Fix wxTextCtrl::SetValue() under wxQt

It shouldn't change the insertion point if the value doesn't really change
However, the event should still be emitted in this case as documented.

Commit:
Enable formerly skipped tests under wxQt

Commit:
Fix testing wxSpinCtrl[Double] unsupported features under wxQt

Backwards ranges are not supported under wxQt and setting empty string
in the control automatically displays minVal.

Commit:
Use cursor type guaranteed to be supported on all platforms.

The wxCURSOR_CHAR cursor type is not supported under wxQt and using it
causes the test to fail there.

Commit:
Fix wxTextCtrl::GetBestSize test case failing under wxQt

Commit:
Fix wxListCtrl test failing under wxQt

The wxListCtrl would assert with this message: '../src/qt/listctrl.cpp(452):
assert "col < columnCount()" failed in SetItem(): Invalid col'

Because under wxQt: no columns means no items, and no items means
no selection can be made at all.

Commit:
Fix windowtest test failing under wxQt

Commit:
Fix gridtest test failing under wxQt

Commit:
Avoid Qt warnings about inactive painter

Qt spits out annoying warnings about inactive painter (QPainter::font:
Painter not active) when using wxDC::GetChar{Width,Height}(). Simply mute
them by using the default application font explicitly.

Commit:
Fix listboxtest test failing under wxQt

Qt uses QString::localeAwareCompare() for comparison and this is what their
documentation says: "On other Unix-like systems without ICU, the comparison
falls back to the system library's strcoll()". Under MSW, Qt uses CompareStringEx()
IOW, Qt always compares strings by locale-specific sorting criteria and BTW
this is what wxGTK was like before this commit 625ed74 (Make sorting of wxListBox,
wxChoice and wxComboBox identical to the MSW sorting)

Commit:
Fix inserting new items into sorted wxListBox under wxQt

Commit:
Fix inserting new items into wxListBox under wxQt

Use the right type when inserting new items into the list box. This ensures that
the wxEVT_CHECKLISTBOX event is emitted for these new items. And in fact, it should
be part of this commit 16da1a1 when wxQtListWidgetItem was introduced.

Commit:
Skip test which work locally, but not when run on GitHub CI under wxQt

Commit:
Fix GetFilename[s] returning extensionless filename[s] under wxQt

Commit:
Fix regular expression used in wxQtAction::Normalize()

Added in commit a605ee3 (Don't pass hyphenated shortcuts to Qt) which
mistakenly replaces the last minus sign (-) of an accelerator with plus
sign (+).

Commit:
Fix wxImageList test failing under wxQt

Commit:
Don't return invalid pixel size from GetPixelSize() under wxQt

AliKet commented:
@vadz This [05bf4ee](https://github.com/wxWidgets/wxWidgets/pull/23925/commits/05bf4ee6043f98157cf2e6e7a8e587d3ab206fd2)  is just a temporary commit to show that the generic `wxPalette` implementation works perfectly under `wxQt`. I need bakefile 0.2.13 to make a real commit. I tried to build it from source but no luck sorry!

vadz commented:
> @vadz This [05bf4ee](https://github.com/wxWidgets/wxWidgets/pull/23925/commits/05bf4ee6043f98157cf2e6e7a8e587d3ab206fd2) is just a temporary commit to show that the generic `wxPalette` implementation works perfectly under `wxQt`.

So should this commit be reverted before merging?

Commit:
Use the generic implementation of wxPalette under wxQt

Until now, wxPalette under wxQt was just a dummy implementation
Let's turn to the generic one which works perfectly under wxQt too.

AliKet commented:
> > @vadz This [05bf4ee](https://github.com/wxWidgets/wxWidgets/pull/23925/commits/05bf4ee6043f98157cf2e6e7a8e587d3ab206fd2) is just a temporary commit to show that the generic `wxPalette` implementation works perfectly under `wxQt`.
> 
> So should this commit be reverted before merging?

I managed to build bakefile 0.2.13 from source (after manually disabling documentation building which was problematic for me) and pushed the updated commit. You can merge this after GitHub CI completed. TIA!

AliKet commented:
@vadz BTW do you know how to enable these tests under AppVeyor ? what the instructions needed to do that ? TIA.

vadz commented:
This seems to be determined by https://github.com/wxWidgets/wxWidgets/blob/368a13a1a2b0d8f9a7c31935cb22dc1537e86d0b/build/tools/appveyor-test.bat#L50 so it looks like it might actually be enabled? Anyhow, this is what is being used by `appveyor.yml`.

vadz commented on interface/wx/spinctrl.h:
```diff
@@ -136,6 +136,8 @@ class wxSpinCtrl : public wxControl
     /**
         Returns the text in the text entry part of the control.
 
+        @note Always returns non-empty string under wxQt.
```
Interesting, so what does it return if the text control is actually empty?

vadz commented on interface/wx/spinctrl.h:
```diff
@@ -189,6 +191,8 @@ class wxSpinCtrl : public wxControl
         @note Setting a range including negative values is silently ignored
         if current base is set to 16.
+
+        @note Backwards ranges are not supported under wxQt.
```
I'd make it more prescriptive:
```suggestion
        @note In wxQt @a minVal must be less than @a maxVal.
```

vadz commented on src/qt/textentry.cpp:
```diff
@@ -104,11 +105,20 @@ wxString wxTextEntry::DoGetValue() const
 }
 
-void wxTextEntry::DoSetValue(const wxString &WXUNUSED(value), int WXUNUSED(flags))
+void wxTextEntry::DoSetValue(const wxString& value, int flags)
 {
+    wxTextEntryBase::DoSetValue(value, flags);
 }
```
Maybe we should just stop overriding it? Or do you plan to do something else than calling the base class version here in the (near) future?

AliKet replied:
No I don't have any plan about this function. at least for now. So feel free to remove this override when you merge this PR. Thanks!

vadz commented on src/qt/window.cpp:
```diff
@@ -350,6 +350,8 @@ bool wxWindowQt::Create( wxWindowQt * parent, wxWindowID id, const wxPoint & pos
     // that a generic control, like wxPanel, is being created, so we need a very
     // simple control as a base:
 
+    wxSize size_ = size;
```
Minor, but `size_` is not very readable. Let's rename it to `initialSize` maybe?

vadz commented on tests/controls/listboxtest.cpp:
```diff
@@ -109,12 +109,21 @@ void ListBoxTestCase::Sort()
 
     m_list->Append(testitems);
 
+#ifndef __WXQT__
```
The different sort order should ideally be documented (I assume we can't change it).

AliKet replied:
we can change the default sort order by overriding the  `operator <` :
```
    virtual bool wxQtListWidgetItem::operator<(const QListWidgetItem& other) const override
    {
        const auto str1 = text();
        const auto str2 = other.text();
        return str1.compare(str2, Qt::CaseInsensitive) < 0;
    }
```
Sure it will make the test happy then (as it passes successfully with it) but how about the end user ? would the list be sorted the way he/she wants/expects ? because the above code doesn't take into account the (locale-specific) rules on how to compare  strings!
IMHO `wxQt` and `wxGTK` (before this commit https://github.com/wxWidgets/wxWidgets/commit/625ed743015f2da84db47a6851e6011bd08f4bd2) looks more correct to me because they respect the `LC_COLLATE` setting of the location.

vadz reviewed (commented):
Great, thanks a lot for all your work here!

There are a couple of possible minor improvements/clarifications but otherwise this is ready to be merged AFAICS.

Thanks again!

Commit:
Now wxDatePickerCtrl is a native control under wxQt

Commit:
Now wxTimePickerCtrl is a native control under wxQt

vadz commented on include/wx/qt/datectrl.h:
```diff
@@ -0,0 +1,64 @@
+// Name:        wx/qt/datectrl.h
+// Purpose:     wxDatePickerCtrl for Qt
+// Author:      Ali Kettab
+// Modified by:
```
Not worth updating the PR just for this, but, just FYI, we don't use this line any more in the new files, it doesn't seem worth it with `git shortlog` providing the same information much more reliably.

vadz commented on include/wx/qt/datectrl.h:
```diff
@@ -0,0 +1,64 @@
+{
+public:
+    // ctors
+    wxDatePickerCtrl() { }
```
Again, not worth changing it here, but in the future, consider writing
```suggestion
    wxDatePickerCtrl() = default;
```
instead.

vadz reviewed (commented):
Great, thanks, I'll (finally) merge this soon.
